File: nodejs/app/UserService.js
'use strict';

const UserRepo = require('./UserRepo');

class UserService {

    constructor() {
        this.repo = new UserRepo;
    }

    getAll(filters = {}) {
        return this.repo.getAll(filters);
    }

    getAllUserHasHeatZone(parameters) {

        let filters = {};

        if( parameters.hasOwnProperty('x')) {
            const minx = Number(parameters.x) - Number( parameters.radius || 0 );
            const maxx = Number(parameters.x) + Number( parameters.radius || 0 );
            filters.x = { $gte: minx, $lte: maxx };
        }
        if( parameters.hasOwnProperty('y')) {
            const miny = Number(parameters.y) - Number( parameters.radius || 0 );
            const maxy = Number(parameters.y) + Number( parameters.radius || 0 );
            filters.y = { $gte: miny, $lte: maxy };
        }
        if( parameters.hasOwnProperty('user')) {
            filters.user = parameters.user;
        }

        return this.getAll(filters)
        .then( (heats) => {

            let users = [];
            heats.map((x) => {
                if(users.indexOf(x.user) == -1) {
                    users.push(x.user);
                }
             });

            return users;
        });
    }

    store(data) {
        return this.repo.store(data);
    }
}

module.exports = UserService;

File: golang/app/services/UserService.go
package services

import (
	"app/app/repositories"
	"gopkg.in/mgo.v2/bson"
	"app/app/models"
)

type UserService struct {
	repo repositories.UserRepo
}

func (u *UserService) Init() {
	u.repo = repositories.UserRepo{}
}

func (u *UserService) ShowMany(filters bson.M) ([]models.User) {
	models := u.repo.ShowAll(filters)
	return models
}

func (u *UserService) GetAllUserHasHeatZone(x int, y int, radius int) ([]string) {

	filters := bson.M{}

	minx := x - radius
	maxx := y + radius
	filters["x"] = bson.M{
		"$gte": minx,
		"$lte": maxx,
	}

	miny := x - radius
	maxy := y + radius
	filters["y"] = bson.M{ 
		"$gte": miny,
		"$lte": maxy,
	}

	models := u.ShowMany(filters)
	var users []string

	for _, element := range models {
		if ! u.contains(users,element.User) {
			users = append(users, element.User)
		}
	}

	return users
}

func (u *UserService) contains(sliceValues []string, search string) bool {
    for _, v := range sliceValues {
        if v == search {
            return true
        }
    }
    return false
}

func (u *UserService) Store(body bson.M) (bson.M, error) {
	return u.repo.Store(body)
}